File: python/cote/LTCD/이진트리_균형_여부.py
from __future__ import annotations
from typing import Generic, TypeVar
from collections import deque
import random

TEST_CASES = 10
MAX_VALUE = 100
MAX_DEPTH = 15

T = TypeVar("T")

class Node(Generic[T]) :
	def __init__(self, value : T, left: Node[T] = None, right: Node[T] = None) :
		self.value = value
		self.left = left
		self.right = right

	def __iter__(self) :
		queue = deque([self])

		while queue :
			node = queue.popleft()

			if node :
				yield node.value

				queue.append(node.left)
				queue.append(node.right)
			else :
				yield None

	def __str__(self) :
		return str([*self])


def initialize(depth: int = 0) -> Node[int] :

	if depth >= MAX_DEPTH :
		return None 

	if depth > 0 and random.random() / depth < 0.00005 / MAX_DEPTH :
		return None

	left, right = initialize(depth + 1), initialize(depth + 1)

	return Node(random.randrange(MAX_VALUE), left, right)


def solve(node: Node[T]) -> bool :

	def check(node: Node[T]) -> int :

		if not node :
			return 0 

		left, right = check(node.left), check(node.right)

		if abs(left - right) > 1 or left == -1 or right == -1 :
			return -1

		return max(left, right) + 1


	return check(node) != -1

if __name__ == "__main__" :
	random.seed(43)

	for _ in range(TEST_CASES) :
		node = initialize()
		print(solve(node))

File: javascript/cote/LTCD/이진트리_균형_여부.mjs

const TEST_CASES = 10
const MAX_VALUE = 100
const MAX_DEPTH = 20

class Node {

	constructor(value, left = null, right = null) {
		this.value = value
		this.left = left
		this.right = right
	}

	*[Symbol.iterator]() {
		const queue = [this]

		while (queue.length) {
			const node = queue.shift()

			if (node) {
				yield node.value
				queue.push(node.left)
				queue.push(node.right)
			} else {
				yield null
			}
		}
	}
}

function initialize(depth = 0) {
	if (depth >= MAX_DEPTH) {
		return null
	}

	if (depth > 0 && Math.random() / depth < 0.000002 / MAX_DEPTH) {
		return null
	}

	const value = Math.floor(Math.random() * MAX_VALUE)
	const [left , right] = [initialize(depth + 1), initialize(depth + 1)]

	return new Node(value, left, right)
}

function solve(node) {

	function dfs(cur) {

		if (! cur) {
			return 0
		}

		const [left ,right] = [dfs(cur.left), dfs(cur.right)]
		
		if (left == -1 || right == -1 || Math.abs(left - right) > 1) {
			return -1
		}

		return Math.max(left, right) + 1
	}

	return dfs(node) != -1
}

for (let it = 0; it < TEST_CASES; it++) {
	const node = initialize()
	console.log(solve(node))
}